### PR TITLE
Warp-RNNT and Static/Dynamic build mode

### PR DESCRIPTION
**Pull request by JackTemaki**
Adds the Warp-RNNT implementation from https://github.com/1ytic/warp-rnnt

Supports building via setup.py but also dynamic loading via [torch.util.cpp_extensions.load](https://pytorch.org/docs/stable/cpp_extension.html#torch.utils.cpp_extension.load) for Warp-RNNT. 

Dynamic loading and Autograd op for Fast-Baum-Welch have to be added later.

### CODE REVIEW

File: i6_native_ops/warp_rnnt/__init__.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    __version__ = get_distribution('i6_native_ops').version
+    import i6_native_ops.warp_rnnt.warp_rnnt_core as core
+except Exception as e:
+    # otherwise try to build locally
+    from torch.utils.cpp_extension import load
+    base_path = os.path.dirname(__file__)
+    core = load(
+        name="warp_rnnt_core",
+        sources=[
+                f"{base_path}/core.cu",
```
Comment (DanEnergetics) [inline i6_native_ops/warp_rnnt/__init__.py:17]: Is there a reason not to write:
```suggestion
                os.path.join(base_path, "core.cu"),
```
Also in the following, e.g. also `setup.py`.

**Review by DanEnergetics — COMMENTED** [review]
How does dynamic loading work now? My naive attempt does not work
```
>>> sys.path.insert(0, "/path/to/containing/folder")
>>> import i6_native_ops
[Works]
>>> import i6_native_ops.warp_rnnt as rnnt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'i6_native_ops.warp_rnnt'
```
Do I need to use `from torch.utils.cpp_extension import load` everytime I want to load it?

File: i6_native_ops/warp_rnnt/__init__.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+import os
+import torch
+from typing import Optional, AnyStr, Literal
+from pkg_resources import get_distribution
+
+try:
+    # Package is installed, so ops are already compiled
+    __version__ = get_distribution('i6_native_ops').version
+    import i6_native_ops.warp_rnnt.warp_rnnt_core as core
```
Comment (curufinwe) [inline i6_native_ops/warp_rnnt/__init__.py:9]: Why not a relative import?

File: i6_native_ops/warp_rnnt/__init__.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                f"{base_path}/core.cu",
+                f"{base_path}/core_gather.cu",
+                f"{base_path}/core_compact.cu",
+                f"{base_path}/binding.cpp"
+        ]
+	)
+
+
+class RNNTLoss(torch.autograd.Function):
+
```
Comment (curufinwe) [inline i6_native_ops/warp_rnnt/__init__.py:26]: Please no empty lines at the beginning of new indentation (also in many other places throughout this file).

File: i6_native_ops/warp_rnnt/__init__.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        return grads_input, None, None, None, None, None, None
+
+
+def rnnt_loss(log_probs: torch.FloatTensor,
+              labels: torch.IntTensor,
+              frames_lengths: torch.IntTensor,
+              labels_lengths: torch.IntTensor,
+              average_frames: bool = False,
+              reduction: Literal['sum', 'mean', 'none'] = 'none',
```
Comment (curufinwe) [inline i6_native_ops/warp_rnnt/__init__.py:78]: An enum would be nicer here

File: i6_native_ops/warp_rnnt/binding.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    at::Tensor counts = at::zeros(counts_shape, counts_opts);
+    at::Tensor alphas = at::empty(buffer_shape, buffer_opts);
+    at::Tensor betas = at::empty(buffer_shape, buffer_opts);
+
+    auto stream = c10::cuda::getCurrentCUDAStream(xs.device().index());
+
+    rnntStatus_t status;
+
+    if (blank == -1) {
+
```
Comment (curufinwe) [inline i6_native_ops/warp_rnnt/binding.cpp:82]: Also here empty lines

File: i6_native_ops/warp_rnnt/__init__.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        costs = RNNTLossCompact.apply(
+            log_probs.float(), 
+            labels, frames_lengths,
+            labels_lengths, blank, 
+            fastemit_lambda, 
+            (log_probs.requires_grad and torch.is_grad_enabled())
+        )
+    else:
+        if gather:
+
```
Comment (curufinwe) [inline i6_native_ops/warp_rnnt/__init__.py:135]: Way too many empty lines here

**Review by curufinwe — CHANGES_REQUESTED** [review]
Some style comments for now. Will test if I can compile this next.

**Comment by JackTemaki** [comment]
@curufinwe the code you commented on is the one copied from the original warp_rnnt repo without any modification. Do we really want to edit it?

**Comment by JackTemaki** [comment]
> How does dynamic loading work now? My naive attempt does not work
> 
> ```
> >>> sys.path.insert(0, "/path/to/containing/folder")
> >>> import i6_native_ops
> [Works]
> >>> import i6_native_ops.warp_rnnt as rnnt
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ModuleNotFoundError: No module named 'i6_native_ops.warp_rnnt'
> ```
> 
> Do I need to use `from torch.utils.cpp_extension import load` everytime I want to load it?

The repo now follows the python convention that the package is "within" the repo, so your path insert is probably wrong, it should be:
```sys.path.insert(0, "/path/to/containing/folder/i6_native_ops")```

**Comment by curufinwe** [comment]
> @curufinwe the code you commented on is the one copied from the original warp_rnnt repo without any modification. Do we really want to edit it?

Fine, let's leave it alone to make updating with new code from upstream easier.

**Comment by curufinwe** [comment]
It seems that now to get to the fbw op we need to do:
`from i6_native_ops.fbw.fbw_core import fbw`. @DanEnergetics Was this the case before aswell? If not we should make sure previously written import statements do not break.

**Comment by JackTemaki** [comment]
> It seems that now to get to the fbw op we need to do: `from i6_native_ops.fbw.fbw_core import fbw`. @DanEnergetics Was this the case before aswell? If not we should make sure previously written import statements do not break.

This was not the case, I hope this can be fixed with a symlink. But I also recommend that you also put the PyTorch code for the op in the repo here before starting setups with the new syntax, for me it was surprising to see only the native code but not the torch wrapping code here.

**Comment by DanEnergetics** [comment]
Yes, this would technically break some setups. However, it appears only in one line in the Sisyphus recipes so far so it would be easy to fix. Otherwise the `warp_rnnt` op and the `fbw` op compile for me so I would be okay with merging.

**Comment by JackTemaki** [comment]
@DanEnergetics I added a symlink, can you check if the old import works like this?

**Comment by DanEnergetics** [comment]
The symlink would probably only help the dynamic loading. Duplicating the CUDAExtension with a different name fixes it for me.

**Comment by JackTemaki** [comment]
Ok then I remove the symlink!

**Review by DanEnergetics — APPROVED** [review]

**Comment by curufinwe** [comment]
Do we need to fix this with code duplication? Couldn't we just add an `__init__.py` file and import the name from the new location? Also for the new location I would suggest creating a package `i6_native_ops.fbw` and importing the fbw function from `i6_native_ops.fbw.core`.

**Comment by DanEnergetics** [comment]
> Do we need to fix this with code duplication?

We could, as you said, also make an `__init__.py` file that imports fbw. In order to guarantee that past imports still work, we would have to have this file in a nativeops folder though since the import before was `from nativeops import fbw`. So something like this
```
i6_native_ops
├── i6_native_ops
|   ├── warp_rnnt
|   ├── fbw
├── nativeops
    ├── __init__.py
```
And then just `from i6_native_ops.fbw.fbw_core import fbw` in the new `__init__.py`.
Again, I do not care about keeping the legacy import that much, because so far it was only me using it, I did not push any code to any other repo that contains this and there is only one line I would have to change overall.

> for the new location I would suggest creating a package i6_native_ops.fbw and importing the fbw function from i6_native_ops.fbw.core

This is reasonable but I could do this also in another PR, though. I wanted to add some python code to the fbw package.

**Comment by curufinwe** [comment]
OK, I also talked to @JackTemaki offline about this and we agreed to skip backward compatibility. I have pushed the relevant changes + a new __init__.py for the fbw module. Are you two @DanEnergetics @JackTemaki OK with these changes?

**Comment by DanEnergetics** [comment]
> a new init.py for the fbw module

I can't seem to see it. Otherwise I'm okay with the changes.

**Comment by JackTemaki** [comment]
@curufinwe same as @DanEnergetics  I do not see the changes

**Comment by curufinwe** [comment]
Ups, forgot to add the __init__.py file. Now it's as it should be.

**Pull request merged**